invenio_vcs/models.py

```diff
@@ -0,0 +1,351 @@
+
+    DELETED = "E"
+    """Release has been deleted."""
+
+    def __init__(self, value):
+        """Hack."""
+
+    def __eq__(self, other):
+        """Equality test."""
+        return self.value == other
+
+    def __str__(self):
+        """Return its value."""
+        return self.value
+
+    @property
+    def title(self):
+        """Return human readable title."""
+        return RELEASE_STATUS_TITLES[self.name]
+
+    @property
+    def icon(self):
+        """Font Awesome status icon."""
+        return RELEASE_STATUS_ICON[self.name]
+
+    @property
+    def color(self):
+        """UI status color."""
+        return RELEASE_STATUS_COLOR[self.name]
+
+
+repository_user_association = db.Table(
+    "vcs_repository_users",
+    db.Model.metadata,
+    db.Column(
+        "repository_id",
+        UUIDType,
+        db.ForeignKey("vcs_repositories.id"),
+        primary_key=True,
+    ),
+    db.Column(
+        "user_id", db.Integer, db.ForeignKey("accounts_user.id"), primary_key=True
+    ),
+)
+
+
+class Repository(db.Model, Timestamp):
+    """Information about a vcs repository."""
+
+    __tablename__ = "vcs_repositories"
+
+    __table_args__ = (
+        UniqueConstraint(
+            "provider",
+            "name",
+            name="uq_vcs_repositories_provider_name",
+        ),
+        UniqueConstraint(
+            "provider",
+            "provider_id",
+            name="uq_vcs_repositories_provider_provider_id",
+        ),
+        # Index("ix_vcs_repositories_provider_provider_id", "provider", "provider_id"),
+    )
+
+    id = db.Column(
+        UUIDType,
+        primary_key=True,
+        default=uuid.uuid4,
+    )
+    """Repository identifier."""
+
+    provider_id = db.Column(
+        db.String(255),
+        nullable=False,
+    )
+    """Unique VCS provider identifier for a repository.
+
+    .. note::
+
+        Past implementations of GitHub for Invenio, used the repository name
+        (eg. 'inveniosoftware/invenio-github') in order to track repositories.
+        This however leads to problems, since repository names can change and
+        thus render the stored repository name useless. In order to tackle this
+        issue, the `provider_id` should be used to track repositories, which is a
+        unique identifier that GitHub uses for each repository and doesn't
+        change on renames/transfers.
+
+        In order to be able to keep deleted repositories with releases that
+        have been published, it is possible to keep an entry without a
+        `provider_id`, that only has a `name`. This only applies to the default
+        `github` provider on migrated pre-VCS instances.
+    """
+
+    provider = db.Column(db.String(255), nullable=False)
+    """Which VCS provider the repository is hosted by (and therefore the context in which to consider the provider_id)"""
+
+    description = db.Column(db.String(10000), nullable=True)
+    html_url = db.Column(db.String(10000), nullable=False)
```

Comment:
these are exported from the extra_data column?

Reply:
Thanks, that makes sense! To be fair we are already generating other URLs such as the new file link, so this isn't too different to that. I will implement it as another generic method on the provider factory class

Reply:
I have replaced `html_url` with the `url_for_repository` and `url_for_release` methods in the provider factory (see [here](https://github.com/inveniosoftware/invenio-github/pull/193/files#diff-0d00acbf4d7b3a42d7e312a5b41c367d9b978b1835e1f47fff615685bde49df6R99-R109))